lib/ddtrace/contrib/rack/middlewares.rb

```diff
@@ -0,0 +1,51 @@
+require 'ddtrace/ext/http'
+
+module Datadog
+  module Contrib
+    # Rack module includes all middlewares to
+    module Rack
+      # TraceMiddleware ensures that the Rack Request is properly traced
+      # from the beginning to the end. The middleware sets the created request
+      # span in the Rack environment so that it can be retrieved in the underlying
+      # application. If request tags are not set by the app, they will be set using
+      # information available in the Rack layer
+      class TraceMiddleware
+        def initialize(app, options = {})
+          @app = app
+          @tracer = options.fetch(:tracer, Datadog.tracer)
+          @service = options.fetch(:default_service, 'rack-app')
+        end
+
+        def call(env)
+          # get the current Rack request
+          request = ::Rack::Request.new(env)
+
+          # start a new request span and attach it to the current Rack environment
+          # Note: the resource is not set because Rack doesn't have a built-in
+          # router; because of that, we can't assume here what is the best
+          # representation for this trace, so we attach the span in the Rack env
+          # so that the underlying Rack App can provide and set more details
+          request_span = @tracer.trace('rack.request', service: @service, span_type: Datadog::Ext::HTTP::TYPE)
+          request.env[:datadog_request_span] = request_span
```

Comment:
Nitpicking, but I'm hesitating with something like `:datadog_rack_request_span`, even more explicit. Because here, it's obvious it's a Rack request. But, within a framework (eg Rails) that is accessing this afterwards to update it, `request` has also a local meaning (the rails request). So in the end you have the Rails request updating the meta (resource) for the Rack request, and in the long run, I'd say having `rack` in the name makes it obvious it belongs to the Rack world. But then, `:datadog_rack_request_span` is 26 chars so somewhat long ;) I let you decide on this, just a thought.